packages/x-data-grid-pro/src/hooks/features/serverSideTreeData/utils.ts

```diff
@@ -3,16 +3,20 @@ import { getTreeNodeDescendants } from '@mui/x-data-grid/internals';
 
 export function skipFiltering(rowTree: GridRowTreeConfig) {
   const filteredRowsLookup: Record<GridRowId, boolean> = {};
+  const filteredChildrenCountLookup: Record<GridRowId, number> = {};
   const filteredDescendantCountLookup: Record<GridRowId, number> = {};
 
   const nodes = Object.values(rowTree);
   for (let i = 0; i < nodes.length; i += 1) {
     const node: any = nodes[i];
     filteredRowsLookup[node.id] = true;
+    filteredChildrenCountLookup[node.id] =
+      node.serverChildrenCount !== undefined ? node.serverChildrenCount : -1;
```

Comment:
I am wondering what will happen if a user uses the `getChildrenCount` as:
```tsx
function getChildrenCount(row) {
  if (row.hasChildren) { // for group nodes
    return row.childrenCount;
  }
  return undefined; // for leaf nodes
}
```

Or simply:
```tsx
function getChildrenCount(row) {
  if (row.hasChildren) { // for group nodes
    return row.childrenCount;
  }
}
```

This might not be the best usage as per the types, but we may expect someone to interpret it this way. So falling back to `-1` when `count='undefined'` might make the grid misinterpret this state value. Alternatively, could we expect `-1` to be explicitly passed by the users, to treat the default node as a leaf node (having `0` children), something like: 

```suggestion
      node.serverChildrenCount ?? 0;
```

What do you think?

Reply:
It makes more sense that the fallback is `0` rather than `-1`
But the interface should prevent you from having the functions as you have written as you have to return a `number`
I will update the description in the [model](https://github.com/mui/mui-x/blob/v7.11.0/packages/x-data-grid/src/models/gridDataSource.ts#L77) to include the unknown children part from the [docs](https://github.com/mui/mui-x/blob/v7.11.0/docs/data/data-grid/server-side-data/tree-data.md?plain=1#L12)

Reply:
> But the interface should prevent you from having the functions as you have written as you have to return a number

You are right, but we have javascript only users too, they might not care much about the interface 😅 

> I will update the description in the [model](https://github.com/mui/mui-x/blob/v7.11.0/packages/x-data-grid/src/models/gridDataSource.ts#L77) to include the unknown children part from the [docs](https://github.com/mui/mui-x/blob/v7.11.0/docs/data/data-grid/server-side-data/tree-data.md?plain=1#L12)

👍